src/app/app.component.ts

```diff
@@ -0,0 +1,100 @@
+import { Component, OnDestroy, OnInit } from '@angular/core';
+import {
+  combineLatest,
+  filter,
+  forkJoin,
+  map,
+  Observable,
+  Subject,
+  Subscription,
+  debounceTime,
+  switchMap,
+} from 'rxjs';
+import { MockDataService } from './mock-data.service';
+
+@Component({
+  selector: 'my-app',
+  templateUrl: './app.component.html',
+  styleUrls: ['./app.component.css'],
+})
+export class AppComponent implements OnInit, OnDestroy {
+  searchTermByCharacters = new Subject<string>();
+  charactersResults$: Observable<any>;
+  planetAndCharactersResults$: Observable<any>;
+  isLoading: boolean = false;
+
+  constructor(private mockDataService: MockDataService) {}
+
+  ngOnInit(): void {
+    this.initLoadingState();
+    this.initCharacterEvents();
+  }
+
+  changeCharactersInput(element): void {
+    // 1.1. Add functionality to changeCharactersInput method. Changes searchTermByCharacters Subject value on input change.
+
+    const inputValue: string = element.target.value;
+    this.searchTermByCharacters.next(inputValue);
+  }
+
+  initCharacterEvents(): void {
+    // 1.2. Add API call on each user input. Use mockDataService.getCharacters - to make get request.
+
+    // 2. Since we don't want to spam our service add filter by input value and do not call API until a user enters at least 3 chars.
+
+    // 3. Add debounce to prevent API calls until user stop typing.
+
+    this.charactersResults$ = this.searchTermByCharacters
```

Comment:
you can avoid creating million of methods and their chains just to create an observable. You can simply define it as a class property.

```suggestion
    export class AppComponent {
      private readonly charactersResults$: Observable<any> = 
         this.searchTermByCharacters.pipe(
            map(characters => (characters ? characters.trim() : "")),
            filter(inputValue => inputValue.length >= 3),
            debounceTime(500),
            switchMap(inputValue => this.mockDataService.getCharacters(inputValue)),
        );
    }
```

In that case it would be much easier because you'll have to write much less code.

Also it's always better to avoid `any` type at any cost, because it neglects typescript as a tool. Using any means no need in typescript, from my POV

Reply:
moved to a class property 

about any:
this code was written in the original file, the task wasn't about fixing types but writing methods
please, correct me if I'm wrong

Reply:
no worries, I got it